### PR TITLE
PROVES-97 Implement sorting of results

### PR DESCRIPTION
**Pull request by kehh**
Fallback on -s.keyword field for sorting text fields
Repeat search without sorting if there's been a problem with searching / sorting

### CODE REVIEW

**Comment by kehh** [comment]
Actually this fails on a reindex:

Index...0.0% 0/244 ETC: ???. Elapsed: < 1 sec [>                              ][07-May-2024 00:42:31 Australia/Melbourne] [2024-05-07 00:42:31] elasticsearch.ERROR: 1 out of 251 bulk operation(s) failed. Errors: Update error: {"type":"document_parsing_exception","reason":"[1:1393] failed to parse field [ca_objects\/date_value.-dtr] of type [date_range] in document with id '1'. Preview of field's value: ''","caused_by":{"type":"document_parsing_exception","reason":"[1:1393] error parsing field [ca_objects\/date_value.-dtr], expected an object but got -dtr"}}. [] []
[07-May-2024 00:42:31 Australia/Melbourne] 1 out of 251 bulk operation(s) failed. Errors: Update error: {"type":"document_parsing_exception","reason":"[1:1393] failed to parse field [ca_objects\/date_value.-dtr] of type [date_range] in document with id '1'. Preview of field's value: ''","caused_by":{"type":"document_parsing_exception","reason":"[1:1393] error parsing field [ca_objects\/date_value.-dtr], expected an object but got -dtr"}}.
[07-May-2024 00:42:31 Australia/Melbourne] [2024-05-07 00:42:31] elasticsearch.ERROR: 1 out of 251 bulk operation(s) failed. Errors: Update error: {"type":"document_parsing_exception","reason":"[1:1393] failed to parse field [ca_objects\/date_value.-dtr] of type [date_range] in document with id '1'. Preview of field's value: ''","caused_by":{"type":"document_parsing_exception","reason":"[1:1393] error parsing field [ca_objects\/date_value.-dtr], expected an object but got -dtr"}}. [] []
[07-May-2024 00:42:31 Australia/Melbourne] 1 out of 251 bulk operation(s) failed. Errors: Update error: {"type":"document_parsing_exception","reason":"[1:1393] failed to parse field [ca_objects\/date_value.-dtr] of type [date_range] in document with id '1'. Preview of field's value: ''","caused_by":{"type":"document_parsing_exception","reason":"[1:1393] error parsing field [ca_objects\/date_value.-dtr], expected an object but got -dtr"}}.
[07-May-2024 00:42:31 Australia/Melbourne] PHP Fatal error:  Uncaught ApplicationException: 1 out of 251 bulk operation(s) failed. Errors: Update error: {"type":"document_parsing_exception","reason":"[1:1393] failed to parse field [ca_objects\/date_value.-dtr] of type [date_range] in document with id '1'. Preview of field's value: ''","caused_by":{"type":"document_parsing_exception","reason":"[1:1393] error parsing field [ca_objects\/date_value.-dtr], expected an object but got -dtr"}}. in /var/www/html/vendor/collectiveaccess/providence/app/lib/Plugins/SearchEngine/Elastic8.php:645
Stack trace:
#0 /var/www/html/vendor/collectiveaccess/providence/app/lib/Plugins/SearchEngine/Elastic8.php(775): WLPlugSearchEngineElastic8->flushContentBuffer()
#1 /var/www/html/vendor/collectiveaccess/providence/app/lib/Plugins/SearchEngine/Elastic8.php(485): WLPlugSearchEngineElastic8->flushBufferWhenFull()
#2 /var/www/html/vendor/collectiveaccess/providence/app/lib/Search/SearchIndexer.php(1155): WLPlugSearchEngineElastic8->commitRowIndexing()
#3 /var/www/html/vendor/collectiveaccess/providence/app/lib/Search/SearchIndexer.php(265): SearchIndexer->indexRow()
#4 /var/www/html/vendor/collectiveaccess/providence/app/lib/Utils/CLIUtils/Search.php(49): SearchIndexer->reindex()
#5 /var/www/html/vendor/collectiveaccess/providence/support/bin/caUtils(171): CLIUtils::rebuild_search_index()
#6 /var/www/html/vendor/bin/caUtils(115): include('/var/www/html/v...')
#7 {main}

Next ApplicationException: 1 out of 251 bulk operation(s) failed. Errors: Update error: {"type":"document_parsing_exception","reason":"[1:1393] failed to parse field [ca_objects\/date_value.-dtr] of type [date_range] in document with id '1'. Preview of field's value: ''","caused_by":{"type":"document_parsing_exception","reason":"[1:1393] error parsing field [ca_objects\/date_value.-dtr], expected an object but got -dtr"}}. in /var/www/html/vendor/collectiveaccess/providence/app/lib/Plugins/SearchEngine/Elastic8.php:645
Stack trace:
#0 /var/www/html/vendor/collectiveaccess/providence/app/lib/Plugins/SearchEngine/Elastic8.php(276): WLPlugSearchEngineElastic8->flushContentBuffer()
#1 /var/www/html/vendor/collectiveaccess/providence/support/bin/caUtils(171): WLPlugSearchEngineElastic8->__destruct()
#2 /var/www/html/vendor/bin/caUtils(115): include('/var/www/html/v...')
#3 {main}
thrown in /var/www/html/vendor/collectiveaccess/providence/app/lib/Plugins/SearchEngine/Elastic8.php on line 645

**Comment by kehh** [comment]
OK @bri-gaia this is good to go

File: app/lib/Plugins/SearchEngine/Elastic8/FieldTypes/DateRange.php [hunk 1]
```diff
@@ -154,4 +154,8 @@ function getFiltersForTerm($term, ?string $field = null): array {
 		return $return;
 	}
 
+	public function getDefaultSuffix(): string {
```
Comment (bri-gaia) [inline app/lib/Plugins/SearchEngine/Elastic8/FieldTypes/DateRange.php]: @kehh this is the line causing the error while indexing. We can remove this, or update parseElasticsearchDateRange() to hardcode the text suffix for the string representation of the date (or remove, unsure if it is needed)

```
public function parseElasticsearchDateRange($content, $historic = false): array {
		$return = [];
		$return[$this->getDataTypeSuffix(FieldType::SUFFIX_TEXT)] = $content;
		...
	}
		```

Comment (kehh) [reply app/lib/Plugins/SearchEngine/Elastic8/FieldTypes/DateRange.php]: Haha sorry you went down that rabbit hole too ;)

**Pull request merged**